pkg/monitoring/endpoint.go

```diff
@@ -0,0 +1,389 @@
+		return
+	}
+}
+
+func setLongerPrefixes(r *http.Request, command *[]string) error {
+	longerPrefixes := r.URL.Query().Get("longer_prefixes")
+	if longerPrefixes != "" {
+		useLongerPrefixes, err := strconv.ParseBool(longerPrefixes)
+		if err != nil {
+			return fmt.Errorf("longer_prefixes value '%s' is not valid: %w", longerPrefixes, err)
+		}
+		if useLongerPrefixes {
+			*command = append(*command, "longer-prefixes")
+		}
+	}
+	return nil
+}
+
+func setInput(r *http.Request, command *[]string) error {
+	input := r.URL.Query().Get("input")
+	if input != "" {
+		if _, _, err := net.ParseCIDR(input); err != nil {
+			return fmt.Errorf("input '%s' is not valid: %w", input, err)
+		}
+		*command = append(*command, input)
+	}
+	return nil
+}
+
+func withNodename(data *[]byte) (*[]byte, error) {
+	res := map[string]json.RawMessage{}
+	nodename := os.Getenv(healthcheck.NodenameEnv)
+
+	result := data
+	if nodename != "" {
+		res[nodename] = json.RawMessage(*data)
+		var err error
+		*result, err = json.MarshalIndent(res, "", "\t")
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling data: %w", err)
+		}
+	}
+
+	return result, nil
+}
+
+func passRequest(r *http.Request, addr, query string, results chan []byte, errors chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	s := strings.Split(r.Host, ":")
+	port := ""
+	if len(s) > 1 {
+		port = s[1]
+	}
+
+	protocol := "http"
+	if r.TLS != nil {
+		protocol = "https"
+	}
+
+	url := fmt.Sprintf("%s://%s:%s%s", protocol, addr, port, query)
+	resp, err := http.Get(url) //nolint
+	if err != nil {
+		errors <- fmt.Errorf("error getting data from %s: %w", addr, err)
+		return
+	}
+
+	data, err := io.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		errors <- fmt.Errorf("error reading response from %s: %w", addr, err)
+		return
+	}
+
+	results <- data
+}
+
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=list
+
+func (e *Endpoint) getAddresses(ctx context.Context, svc *corev1.Service) ([]string, error) {
+	var serviceLabels labels.Set = svc.Spec.Selector
+	pods := &corev1.PodList{}
+	if err := e.c.List(ctx, pods, &client.ListOptions{
+		LabelSelector: serviceLabels.AsSelector(),
+		Namespace:     svc.Namespace,
+	}); err != nil {
+		return nil, fmt.Errorf("error getting pods: %w", err)
+	}
+
+	addresses := []string{}
+	for i := range pods.Items {
+		addresses = append(addresses, pods.Items[i].Status.PodIP)
+	}
+
+	return addresses, nil
+}
+
+func queryEndpoints(r *http.Request, addr []string) ([]byte, error) {
+	query := strings.ReplaceAll(r.URL.RequestURI(), "all/", "")
+	responses := []json.RawMessage{}
+
+	var wg sync.WaitGroup
+	results := make(chan []byte, len(addr))
+	errors := make(chan error, len(addr))
+
+	for i := range addr {
+		wg.Add(1)
+		go passRequest(r, addr[i], query, results, errors, &wg)
+	}
+
+	wg.Wait()
+	close(results)
+	close(errors)
+
+	for err := range errors {
+		return nil, fmt.Errorf("error occurred: %w", err)
+	}
```

Comment:
This will only return the first error, so `if len(errors) > 0 { return nil, errors[0] }` would be more efficient. Always a bit tricky to do this properly, maybe just return a list of errors for now.

In general I've been thinking it might make sense to add proper logging, and when doing that hide a few more errors in the logs instead of returning them in the response.

Reply:
Changed that. If multiple errors will occur, the response will be
```
http.Error(w, "multiple errors occurred while querying endpoints - please check logs for the details", http.StatusInternalServerError)
```

Also added some logging, which looks like this for now:

```
2024-04-18T14:51:06Z	INFO	setup	loaded kubernetes config
2024-04-18T14:51:06Z	INFO	setup	loaded status service config
2024-04-18T14:51:06Z	INFO	monitoring	created ServeMux
2024-04-18T14:51:06Z	INFO	setup	created server, starting...	{"Addr": ":7083", "ReadHeaderTimeout": "20s", "ReadTimeout": "1m0s"}
2024-04-18T14:51:24Z	INFO	monitoring	got ShowRoute request
2024-04-18T14:51:24Z	INFO	monitoring	command to be executed	{"command": ["show", "ip", "route", "vrf", "all"]}
2024-04-18T14:51:25Z	INFO	monitoring	response written	{"type": "ShowRoute"}
2024-04-18T14:51:42Z	INFO	monitoring	got QueryAll request
2024-04-18T14:51:42Z	INFO	monitoring	will querry endpoints	{"endpoints": ["172.18.0.3", "172.18.0.6", "172.18.0.8", "172.18.0.7", "172.18.0.5", "172.18.0.2"]}
2024-04-18T14:51:42Z	INFO	monitoring	got ShowRoute request
2024-04-18T14:51:42Z	INFO	monitoring	command to be executed	{"command": ["show", "ip", "route", "vrf", "all"]}
2024-04-18T14:51:42Z	INFO	monitoring	response written	{"type": "ShowRoute"}
2024-04-18T14:51:42Z	INFO	monitoring	response written	{"type": "QueryAll"}
```